risc0/zkvm/platform/src/rust_rt.rs

```diff
@@ -80,23 +77,54 @@ mod entrypoint {
     );
 }
 
-struct BumpPointerAlloc;
+pub mod heap {
+    #[cfg(not(feature = "heap-embedded-alloc"))]
+    pub mod bump_allocator {
+        use crate::syscall::sys_alloc_aligned;
+        use core::alloc::{GlobalAlloc, Layout};
 
-unsafe impl GlobalAlloc for BumpPointerAlloc {
-    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        sys_alloc_aligned(layout.size(), layout.align())
-    }
+        pub struct BumpPointerAlloc;
+
+        unsafe impl GlobalAlloc for BumpPointerAlloc {
+            unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+                sys_alloc_aligned(layout.size(), layout.align())
+            }
+
+            unsafe fn dealloc(&self, _: *mut u8, _: Layout) {
+                // this allocator never deallocates memory
+            }
+
+            unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+                // NOTE: This is safe to avoid zeroing allocated bytes, as the bump allocator does not
+                //       re-use memory and the zkVM memory is zero-initialized.
+                self.alloc(layout)
+            }
+        }
 
-    unsafe fn dealloc(&self, _: *mut u8, _: Layout) {
-        // this allocator never deallocates memory
+        #[global_allocator]
+        pub static HEAP: BumpPointerAlloc = BumpPointerAlloc;
     }
 
-    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
-        // NOTE: This is safe to avoid zeroing allocated bytes, as the bump allocator does not
-        //       re-use memory and the zkVM memory is zero-initialized.
-        self.alloc(layout)
+    #[cfg(feature = "heap-embedded-alloc")]
+    pub mod embedded_allocator {
+        use critical_section::RawRestoreState;
+        use embedded_alloc::Heap;
+
+        struct MyCriticalSection;
+
+        critical_section::set_impl!(MyCriticalSection);
+
+        unsafe impl critical_section::Impl for MyCriticalSection {
```

Comment:
What is the purpose of this? It doesn't seem to be in use.

Reply:
this is required by the embedded allocator - it needs to know how to lock/unlock to get exclusive access to the heap so that it doesn't run into race conditions